Using associated constant as array length: without (funny error) and with (ICE) const generics

Given the following test and nightly 2020-01-12:
```
#![allow(unused)]
#![feature(const_generics)]

trait Child {
    const N: usize;
}

struct C1;
impl Child for C1 {
    const N: usize = 2;
}

struct Main<C: Child> {
    c: C
}

impl<C: Child> Main<C> {
    fn test() -> [u8; C::N] {
        [0; C::N]
    }
}
```

I get that this is not supposed to work without const generics, but the error message suggestion is very funny:
```
error[E0599]: no associated item named `N` found for type parameter `C` in the current scope
  --> src/main.rs:22:27
   |
22 |     fn test2() -> [u8; C::N] {
   |                           ^ associated item not found in `C`
   |
   = help: items from traits can only be used if the type parameter is bounded by the trait
help: the following trait defines an item `N`, perhaps you need to restrict type parameter `C` with it:
   |
17 | impl<C: Child + Child> Main<C> {
   |      ^^^^^^^^^^
```

With const generics:
```
warning: Error finalizing incremental compilation session directory `/home/gbr/devel/rust-test/target/debug/incremental/rust_test-xwudz02f6bay/s-fjpbjlpyxj-1ihvwt0-working`: No such file or directory (os error 2)

error: internal compiler error: cat_expr Errd
  --> src/main.rs:22:30
   |
22 |       fn test2() -> [u8; C::N] {
   |  ______________________________^
23 | |         [0; C::N]
24 | |     }
   | |_____^

error: internal compiler error: cat_expr Errd
  --> src/main.rs:23:9
   |
23 |         [0; C::N]
   |         ^^^^^^^^^

error: internal compiler error: PromoteTemps: MIR had errors
  --> src/main.rs:22:5
   |
22 | /     fn test2() -> [u8; C::N] {
23 | |         [0; C::N]
24 | |     }
   | |_____^

error: internal compiler error: broken MIR in DefId(0:16 ~ rust_test[3df1]::{{impl}}[1]::test2[0]) ("return type"): bad type [type error]
  --> src/main.rs:22:5
   |
22 | /     fn test2() -> [u8; C::N] {
23 | |         [0; C::N]
24 | |     }
   | |_____^

error: internal compiler error: broken MIR in DefId(0:16 ~ rust_test[3df1]::{{impl}}[1]::test2[0]) (LocalDecl { mutability: Mut, local_info: Other, internal: false, is_block_tail: None, ty: [type error], user_ty: UserTypeProjections { contents: [] }, source_info: SourceInfo { span: src/main.rs:22:5: 24:6, scope: scope[0] } }): bad type [type error]
  --> src/main.rs:22:5
   |
22 | /     fn test2() -> [u8; C::N] {
23 | |         [0; C::N]
24 | |     }
   | |_____^

thread 'rustc' panicked at 'no errors encountered even though `delay_span_bug` issued', src/librustc_errors/lib.rs:346:17
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.42.0-nightly (3ebcfa145 2020-01-12) running on x86_64-unknown-linux-gnu

note: compiler flags: -C debuginfo=2 -C incremental --crate-type bin

note: some of the compiler flags provided by cargo are hidden
```

--- a/src/librustc_typeck/check/method/suggest.rs
+++ b/src/librustc_typeck/check/method/suggest.rs
@@ -853,26 +853,30 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                                 } else {
                                     sp
                                 };
-                                // FIXME: contrast `t.def_id` against `param.bounds` to not suggest
-                                // traits already there. That can happen when the cause is that
-                                // we're in a const scope or associated function used as a method.
-                                err.span_suggestions(
-                                    sp,
-                                    &message(format!(
-                                        "restrict type parameter `{}` with",
-                                        param.name.ident(),
-                                    )),
-                                    candidates.iter().map(|t| {
-                                        format!(
-                                            "{}{} {}{}",
+                                let trait_def_ids: FxHashSet<DefId> = param
+                                    .bounds
+                                    .iter()
+                                    .filter_map(|bound| bound.trait_def_id())
+                                    .collect();
+                                if !candidates.iter().any(|t| trait_def_ids.contains(&t.def_id)) {
+                                    err.span_suggestions(
+                                        sp,
+                                        &message(format!(
+                                            "restrict type parameter `{}` with",
                                             param.name.ident(),
-                                            if impl_trait { " +" } else { ":" },
-                                            self.tcx.def_path_str(t.def_id),
-                                            if has_bounds.is_some() { " + " } else { "" },
-                                        )
-                                    }),
-                                    Applicability::MaybeIncorrect,
-                                );
+                                        )),
+                                        candidates.iter().map(|t| {
+                                            format!(
+                                                "{}{} {}{}",
+                                                param.name.ident(),
+                                                if impl_trait { " +" } else { ":" },
+                                                self.tcx.def_path_str(t.def_id),
+                                                if has_bounds.is_some() { " + " } else { "" },
+                                            )
+                                        }),
+                                        Applicability::MaybeIncorrect,
+                                    );
+                                }
                                 suggested = true;
                             }
                             Node::Item(hir::Item {

--- a/src/test/ui/associated-item/associated-item-duplicate-bounds.rs
+++ b/src/test/ui/associated-item/associated-item-duplicate-bounds.rs
@@ -1,0 +1,11 @@
+trait Adapter {
+    const LINKS: usize;
+}
+
+struct Foo<A: Adapter> {
+    adapter: A,
+    links: [u32; A::LINKS], // Shouldn't suggest bounds already there.
+    //~^ ERROR: no associated item named `LINKS` found
+}
+
+fn main() {}

--- a/src/test/ui/associated-item/associated-item-duplicate-bounds.stderr
+++ b/src/test/ui/associated-item/associated-item-duplicate-bounds.stderr
@@ -1,0 +1,11 @@
+error[E0599]: no associated item named `LINKS` found for type parameter `A` in the current scope
+  --> $DIR/associated-item-duplicate-bounds.rs:7:21
+   |
+LL |     links: [u32; A::LINKS], // Shouldn't suggest bounds already there.
+   |                     ^^^^^ associated item not found in `A`
+   |
+   = help: items from traits can only be used if the type parameter is bounded by the trait
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0599`.

--- a/src/test/ui/issues/issue-39559.stderr
+++ b/src/test/ui/issues/issue-39559.stderr
@@ -5,10 +5,6 @@ LL |     entries: [T; D::dim()],
    |                     ^^^ function or associated item not found in `D`
    |
    = help: items from traits can only be used if the type parameter is bounded by the trait
-help: the following trait defines an item `dim`, perhaps you need to restrict type parameter `D` with it:
-   |
-LL | pub struct Vector<T, D: Dim + Dim> {
-   |                      ^^^^^^^^
 
 error: aborting due to previous error
 

--- a/src/test/ui/span/issue-7575.stderr
+++ b/src/test/ui/span/issue-7575.stderr
@@ -61,7 +61,10 @@ error[E0599]: no method named `is_str` found for type parameter `T` in the curre
   --> $DIR/issue-7575.rs:70:7
    |
 LL |     t.is_str()
-   |       ^^^^^^ this is an associated function, not a method
+   |     --^^^^^^--
+   |     | |
+   |     | this is an associated function, not a method
+   |     help: disambiguate the method call for the candidate: `ManyImplTrait::is_str(t)`
    |
    = note: found the following associated functions; to be used as methods, functions must have a `self` parameter
 note: the candidate is defined in the trait `ManyImplTrait`
@@ -70,14 +73,6 @@ note: the candidate is defined in the trait `ManyImplTrait`
 LL |     fn is_str() -> bool {
    |     ^^^^^^^^^^^^^^^^^^^
    = help: items from traits can only be used if the type parameter is bounded by the trait
-help: disambiguate the method call for the candidate
-   |
-LL |     ManyImplTrait::is_str(t)
-   |
-help: the following trait defines an item `is_str`, perhaps you need to restrict type parameter `T` with it:
-   |
-LL | fn param_bound<T: ManyImplTrait + ManyImplTrait>(t: T) -> bool {
-   |                ^^^^^^^^^^^^^^^^^^
 
 error: aborting due to 3 previous errors
 

